From 1b5d77ab17c80a62a47e84adca6dcf4d67a10bd6 Mon Sep 17 00:00:00 2001
From: WofWca <wofwca@protonmail.com>
Date: Tue, 5 May 2026 22:12:12 +0400
Subject: [PATCH] feat: CVAR to apply collision dmg to live players

This affects gameplay significantly,
and is basically just for fooling around.
---
 code/game/g_active.c | 10 +++++++---
 code/game/g_local.h  |  1 +
 code/game/g_main.c   |  3 +++
 3 files changed, 11 insertions(+), 3 deletions(-)

diff --git a/code/game/g_active.c b/code/game/g_active.c
index 457fbb51..e45392a7 100644
--- a/code/game/g_active.c
+++ b/code/game/g_active.c
@@ -307,8 +307,12 @@ static void CheckCollisionDamage( gentity_t *ent, const vec3_t oldVelocity ) {
 	vec3_t		velChange;
 	int			damage;
 	float		div;
+	float		baseDamage = g_gibsOnCollisionBaseDamage.value *
+		(client->ps.pm_type == PM_DEAD
+			? 1
+			: g_gibsOnCollisionAffectLivePlayers.value);
 
-	if ( client->ps.pm_type != PM_DEAD ) {
+	if ( baseDamage == 0 ) {
 		return;
 	}
 
@@ -322,12 +326,12 @@ static void CheckCollisionDamage( gentity_t *ent, const vec3_t oldVelocity ) {
 	}
 	damage = VectorLengthSquared( velChange )
 		/ div
-		* g_gibsOnCollisionBaseDamage.value;
+		* baseDamage;
 
 	// If the player already has -39 health,
 	// gibbing from a 2-centimeter fall would not look good.
 	// So we gotta have a threshold.
-	if ( damage < g_gibsOnCollisionBaseDamage.integer ) {
+	if ( damage < baseDamage ) {
 		return;
 	}
 
diff --git a/code/game/g_local.h b/code/game/g_local.h
index 569b5d42..eb54ef3b 100644
--- a/code/game/g_local.h
+++ b/code/game/g_local.h
@@ -752,6 +752,7 @@ extern	vmCvar_t	g_gibsNewEvGibPlayerParmProtocol;
 extern	vmCvar_t	g_gibsOnCollisionInheritPlayerVelocity;
 extern	vmCvar_t	g_gibsOnCollisionMinSpeed;
 extern	vmCvar_t	g_gibsOnCollisionBaseDamage;
+extern	vmCvar_t	g_gibsOnCollisionAffectLivePlayers;
 extern	vmCvar_t	g_allowVote;
 extern	vmCvar_t	g_teamAutoJoin;
 extern	vmCvar_t	g_teamForceBalance;
diff --git a/code/game/g_main.c b/code/game/g_main.c
index 05a59bbe..e6542196 100644
--- a/code/game/g_main.c
+++ b/code/game/g_main.c
@@ -75,6 +75,7 @@ vmCvar_t	g_gibsNewEvGibPlayerParmProtocol;
 vmCvar_t	g_gibsOnCollisionInheritPlayerVelocity;
 vmCvar_t	g_gibsOnCollisionMinSpeed;
 vmCvar_t	g_gibsOnCollisionBaseDamage;
+vmCvar_t	g_gibsOnCollisionAffectLivePlayers;
 vmCvar_t	g_podiumDist;
 vmCvar_t	g_podiumDrop;
 vmCvar_t	g_allowVote;
@@ -178,6 +179,8 @@ static cvarTable_t		gameCvarTable[] = {
 	// (damage increases with the square of the speed).
 	{ &g_gibsOnCollisionMinSpeed, "g_gibsOnCollisionMinSpeed", "1000", CVAR_ARCHIVE, 0, qfalse },
 	{ &g_gibsOnCollisionBaseDamage, "g_gibsOnCollisionBaseDamage", "150", CVAR_ARCHIVE, 0, qfalse },
+	// This one affects gameplay significantly, and is basically just for fun.
+	{ &g_gibsOnCollisionAffectLivePlayers, "g_gibsOnCollisionAffectLivePlayers", "0.0", 0, 0, qtrue },
 
 	{ &g_podiumDist, "g_podiumDist", "80", 0, 0, qfalse },
 	{ &g_podiumDrop, "g_podiumDrop", "70", 0, 0, qfalse },
-- 
2.54.0

